storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -978,17 +978,23 @@ fsp_header_fill_encryption_info(
 	/* Write magic header. */
 	if (version == Encryption::ENCRYPTION_VERSION_1) {
 		memcpy(ptr, ENCRYPTION_KEY_MAGIC_V1, ENCRYPTION_MAGIC_SIZE);
-	} else {
+	} else if (version == Encryption::ENCRYPTION_VERSION_2) {
 		memcpy(ptr, ENCRYPTION_KEY_MAGIC_V2, ENCRYPTION_MAGIC_SIZE);
+	} else {
+		memcpy(ptr, ENCRYPTION_KEY_MAGIC_V3, ENCRYPTION_MAGIC_SIZE);
 	}
 	ptr += ENCRYPTION_MAGIC_SIZE;
 
 	/* Write master key id. */
 	mach_write_to_4(ptr, master_key_id);
-	ptr += sizeof(ulint);
+	if (version == Encryption::ENCRYPTION_VERSION_3) {
```

Comment:
While writing, there is no need to write V1 or V2. Just write V3 format (No need check for 'version' during write)

Change the entire patch to
1. Write only V3
2. Handle reading of v1, v2 and v3.

Reply:
I have to keep at least V1: some encryption features are already GA, and V1 stored keys in the keyring under a different name.
For V2, we would lose some (not 100% complete) downgrade ability within 5.7, that's probably not an issue - but then it's only a few lines of code.

Reply:
So tablespaces that come from older 5.7 continue to use V2?  I don't think we should care about downgrade ability before GA.

Also ALTER INSTANCE INNODB ROTATE MASTER KEY should write encryption headers in V3. Please verify this.

Reply:
any rebuild of the tables would bring the tables to V3 format anyway

Reply:
This PR is for 5.7, it is not about the 5.7->8.0 upgrade, it's about how versions within 5.7 behave.

Reply:
@dutow I understand that. So my questions are
1. ALTERs which rebuild table would recreate the tablespace encryption header in V3 format right? please check and confirm
2. ALTER INSTANCE .. on upgraded tables from earlier version. Would they continue using V2 format after rotation?

Reply:
I've added a testcase about this.

Reply:
Thanks, looks good. added comment on that

Reply:
Now your tests confirm that DDLs (and ALTER INSTANCE) will change all tables to V3 and downgrade is not possible.

I will make this suggestion optional. So upto you if you want to write only V3.

Reply:
I'll keep this functionality for now - without it, I wouldn't be able to add this test :)